cdc/processor/sinkmanager/worker_impl.go

```diff
@@ -0,0 +1,170 @@
+			for {
+				e, pos, err := iter.Next()
+				if err != nil {
+					return err
+				}
+				// There is no more data.
+				if e == nil {
+					break
+				}
+				for availableMem-e.Row.ApproximateBytes() < 0 {
+					w.memQuota.ForceAcquire(defaultMemoryUsage)
+					availableMem += defaultMemoryUsage
+				}
+				availableMem -= e.Row.ApproximateBytes()
+				events = append(events, e)
+				// We meet a finished transaction.
+				if pos.Valid() {
+					lastPos = pos
+					if w.splitTxn {
+						w.memQuota.ResetBatchID(t.tableID)
+					}
+					// Always emit the events to the sink.
+					// Whatever splitTxn is true or false, we should emit the events to the sink as soon as possible.
+					size, err := w.emitEventsToTableSink(t, events)
+					if err != nil {
+						return err
+					}
+					lastTotalSize += size
+					events = events[:0]
+					if lastTotalSize >= maxUpdateIntervalSize {
+						err := w.updateTableSinkResolvedTs(t, e.CRTs, currentBarrierTs, lastTotalSize)
+						if err != nil {
+							return err
+						}
+						lastTotalSize = 0
+					}
+					// If we exceed the whole memory quota, we should stop the task.
+					// And just wait for the next round.
+					if w.memQuota.IsExceed() {
+						err := w.updateTableSinkResolvedTs(t, e.CRTs, currentBarrierTs, lastTotalSize)
+						if err != nil {
+							return err
+						}
+						lastTotalSize = 0
+						break
+					}
+				}
+				// If we enable splitTxn, we should emit the events to the sink when the batch size is exceeded.
+				if w.splitTxn && uint64(batchCount) >= w.batchSize {
+					size, err := w.emitEventsToTableSink(t, events)
+					if err != nil {
+						return err
+					}
+					lastTotalSize += size
+					if lastTotalSize >= maxUpdateIntervalSize {
+						err := w.updateTableSinkResolvedTs(t, e.CRTs, currentBarrierTs, lastTotalSize)
+						if err != nil {
+							return err
+						}
+						lastTotalSize = 0
+					}
+					events = events[:0]
+				}
+			}
+			// Do not forget to refund the useless memory quota.
+			w.memQuota.Refund(uint64(availableMem))
+			// Add table back.
+			t.callback(lastPos)
+			if err := iter.Close(); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (w *workerImpl) emitEventsToTableSink(t *tableSinkTask, events []*model.PolymorphicEvent) (uint64, error) {
+	rowChangeEvents := make([]*model.RowChangedEvent, 0, len(events))
+	size := 0
+	for _, e := range events {
+		size += e.Row.ApproximateBytes()
+		rows, err := t.tableSink.verifyAndTrySplitEvent(e)
+		if err != nil {
+			return 0, err
+		}
+		rowChangeEvents = append(rowChangeEvents, rows...)
+	}
+
+	t.tableSink.emitRowChangedEvent(rowChangeEvents...)
+	return uint64(size), nil
+}
+
+func (w *workerImpl) updateTableSinkResolvedTs(t *tableSinkTask, commitTs model.Ts, barrierTs model.Ts, size uint64) error {
+	tableSinkResolvedTs := commitTs
+	if barrierTs < tableSinkResolvedTs {
+		tableSinkResolvedTs = barrierTs
+	}
+	if w.redoManager != nil {
```

Comment:
How about removing redoManager from this function? Let's focus on table sink here. We can handle redo-related things in another function.

And parameter `barrierTs` can be removed, too.

Reply:
Oh. I am stupid. It never over the barrier(aka upper bound).

Reply:
Thanks!